Tiered Dungeons rebalance by Eric

eric-anderson opened the pull request:
### All Submissions:

- [ YES ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Tested Changes:

1. [ YES ] Have you lint your code (lua lint) locally prior to submission?

### Changes to Core Features:

- [ YES ] Have you added an explanation of what your changes do and why you'd like us to include them?

See commit messages. For overall summary of the set of changes see below (where I copied it) or 

https://github.com/eric-anderson/ComfyFactorio/blob/75864b385c38d2cb191d22a3bed207858d7d1179/maps/dungeons/tiered_dungeon.lua#L872

Which is the information for dungeons in the new changelog tab.

- [ YES ] Have you successfully ran tests with your changes locally?

Server is currently running if you want to try it out.

### Comments

	{ ver = '1.1', date = '2022-03-13', desc = [[
* All research is now found at random.
  * Red science floors 0-1
  * Green on floors 1-5
  * Gray on floors 5-10
  * Blue on floors 8-13
  * Blue/gray on floors 10-14
  * Purple on floors 12-19
  * Yellow on floors 14-21
  * White on floors 20-25
  * Atomic Bomb/Spidertron on floors 22-25
* Add melee mode toggle to top bar. Keeps weapons in main inventory if possible.
* Ore from rocks nerfed. Used to hit max value on floor 2, now scales up from
  floors 0-19 along with ore from rooms. After floor 20 ore from rooms scales up faster.
* Treasure rooms
  * Rescaled to have similar total resources regardless of size
  * Unlimited number of rooms but lower frequency
  * Loot is limited to available loot 3 floors lower, but slightly more total value than before.
* Autostash and corpse clearing from Mountain Fortress enabled
* Harder rooms will occur somewhat farther out on the early floors.
* Spawners and worm counts bounded in early rooms.
]]},

eric-anderson committed:
Refactor to move research unlock logic to a separate module

eric-anderson committed:
Bound number of spawners and worms in dirtlands; more gradual difficulty curve

eric-anderson committed:
fix typo

eric-anderson committed:
Make hard rooms (acid/concrete) only appear farther out from the center on upper levels

eric-anderson committed:
Enable randomized technology on all floors

eric-anderson committed:
Enable autostash and corpse clearing lower buttons

eric-anderson committed:
fix nil dereference

eric-anderson committed:
Add support for a button to put character in melee mode so that guns/ammo will not go into weapon/ammo slots

eric-anderson committed:
enable melee mode, move doom rooms under the enable_hard_rooms if test

eric-anderson committed:
add lots of missing locals, fix print to use proper quoting

eric-anderson committed:
tweak recipe unlock pattern to be more controllable. Change research scaling to start hard and scale down as research is found on floors

eric-anderson committed:
maybe fix bool/int error in log with parens, if not, will get better line number next time

eric-anderson committed:
reduce mindist for hard rooms, 200 was too far out. move tech multiplier into research module

eric-anderson committed:
tweak room generation to increase chance of creating room

eric-anderson committed:
relock research that is incorrectly enabled

eric-anderson committed:
Add flag to always allow clearing corpses

eric-anderson committed:
Add hook to enable autostash into the dungeons linked chests

eric-anderson committed:
enable dungeons autostash, enable everyone to clear corpses

eric-anderson committed:
Handle the case where player is dead

eric-anderson committed:
BUGFIX: need to teleport character to dungeon floor before creating character, otherwise creation fails

eric-anderson committed:
Lots more tweaking of the distribution so it is more even across floors. Special case automation so it is always found first on floor 0.

eric-anderson committed:
Add ability to make spaghetti setting a no-op, otherwise the config overrides dungeons settings for research

eric-anderson committed:
Add ability to entirely disable collapse. It is used by RPG and causes log spamming (12G log when noticed) because it can not collapse in tiered dungeon

eric-anderson committed:
Nerf ore from rocks. Old formula hit its maximum on floor 2, new formula hits max on floor 19. Tweak resoruce distribution so less stone is generated on the lower levels and the distribution is closer to the all-science distribution

eric-anderson committed:
Disable collapse and spaghetti mode

eric-anderson committed:
Add changelog module to the panel

eric-anderson committed:
Rebalance ores, noop spaghetti mode, add changelog

eric-anderson committed:
Bugfix: not op binds more tightly than >, so not abs(v) < w becomes (not abs(v)) < w which makes lua sad.
Also fix island to have water all around the outside. After refactor it
became clear the logic was probably missing the second not.

eric-anderson committed:
Fix sundry bugs in the changelog implementation

eric-anderson committed:
Make treasure rooms unbounded but lower frequency the more that appear

eric-anderson committed:
tweak resources in treasure biome down by 2x it was too high

eric-anderson committed:
bugfix: change to have more resources accidentally applied to corridors

eric-anderson committed:
Rebalance item loot so treasure rooms are also limited on available loot

eric-anderson committed:
Merge https://github.com/eric-anderson/ComfyFactorio into master

eric-anderson committed:
Fix issues found by luacheck

eric-anderson committed:
revert change to control.lua

Gerkiz commented:
Made some GUI changes. Moved the files over to .../utils/gui/. Hence the conflict.

hanakocz commented on maps/dungeons/tiered_dungeon.lua:
```diff
@@ -12,8 +12,17 @@ local BiterRaffle = require 'functions.biter_raffle'
+local Autostash = require 'modules.autostash'
+local Panel = require 'comfy_panel.config'
+Panel.get('gui_config').spaghett.noop = true
+local Collapse = require 'modules.collapse'
```
What is the reasoning to include Collapse module there, just to disable it and never use it again?

eric-anderson replied:
modules.rpg.main requires modules.explosives
modules.explosives requires modules.collapse

After about a week of running the server, collapse started trying to collapse tiered dungeons and by the time I noticed I had a 12G log file. So in 
https://github.com/eric-anderson/ComfyFactorio/commit/73b5d03276e8202f62689e7e39d5c934365e94c4

I added the ability to disable collapse entirely and in this commit I disabled collapse.

I didn't feel confident enough in understanding the tree of dependencies in the modules to make a change elsewhere.

eric-anderson replied:
I took another look at it. It looks like if I call Collapse.start_now(false) it will also disable collapse.  It's slightly less safe than the other approach since if dungeons on_init runs before collapse on_init then it won't work right, but based on some manual testing it seems to do the right thing.  Hopefully it won't break after a week.

Gerkiz commented on utils/commands/misc.lua:
```diff
@@ -359,7 +359,7 @@ local function clear_corpses(cmd)
     end
     local p = player.print
-    if not trusted[player.name] then
+    if not trusted[player.name] and not this.always_allow_clear_corpses then
```
It stays trusted only.

eric-anderson replied:
resolved in 26f34ea8d5d108a6cbd20848b24bc02525c5ff98

Gerkiz commented on utils/commands/misc.lua:
```diff
@@ -499,6 +499,10 @@ function Public.reset()
     this.players = {}
 end
 
+function Public.set_always_allow_clear_corpses(val)
```
This can be removed.

eric-anderson replied:
resolved in 26f34ea8d5d108a6cbd20848b24bc02525c5ff98

Gerkiz commented on modules/collapse.lua:
```diff
@@ -270,12 +272,16 @@ local function on_init()
 
 local function on_tick()
-    if game.tick % collapse.speed ~= 0 then
+    if game.tick % collapse.speed ~= 0 or collapse.disabled then
```
Collapse "enabled/disabled" is already checked on the function, so this can be removd.

eric-anderson replied:
Can you tell me where it's checked? As mentioned above my experience was that after a week of the server up it started log spamming about not being able to collapse.

eric-anderson replied:
Removed, I found another way to do this by calling Collapse.start_now(false)

Gerkiz commented on modules/collapse.lua:
```diff
@@ -270,12 +272,16 @@ local function on_init()
     progress()
 end
 
+function Public.disable()
```
Same as line 275.

eric-anderson replied:
Removed.

Gerkiz commented on modules/changelog.lua:
```diff
@@ -0,0 +1,107 @@
+local Event = require 'utils.event'
```
This module does not make much sense, we have map info to list these things.

eric-anderson replied:
I wanted to separate between the changes that were made to a particular map and the information about a map. I can certainly modify it so that I can put a changelog at the end of the map. It seemed nice to let people playing know what had changed since the last version.

Gerkiz replied:
Valid reason, ty!

Gerkiz commented on modules/autostash.lua:
```diff
@@ -58,6 +58,20 @@ local function prepare_floaty_text(list, surface, position, name, count)
 
 local function chest_is_valid(chest)
+   if this.dungeons_initial_level ~= nil then
+      -- game.print('ERIC cpos ' .. serpent.line(chest.position))
```
This does not belong here. Don't leave any comments.

hanakocz replied:
I think that this autostash code is not needed in general, allowing the transport chests as valid containers is probably not the best idea in general (players should have good control what goes into them, and there are loaders available in most of scenarios), but even then, including blue chest entity type in allowed chests would be clever way (even if conditional for one map) instead of checking for coordinates of each scanned chest.

eric-anderson replied:
This approach seemed to match with the way people tended to play "dump loot into the transport chests so it goes up to a sorter"

I can certainly remove it. In previous runs of the map I found myself manually dumping things into the transport chests since there was no easy way to auto load them. I guess the next closest would be have a chest and a loader and always pick up the loader, empty stuff and then put the loader back.

The reason for specific coordinates rather than all blue chests is to avoid the problem of dumping into blue chests used for logistics factories.

hanakocz replied:
Blue Chest is special entity, it only shares graphics with the logistic blue chest. It is creative-only chest (same as red chest, both are used for transporting stuff for that unique reason). So if we do "if dungeons then blue chest is whitelisted", then it is one operation and it will automatically find the chest.

The whitelist still is just first step, the "if chest is used (e.q. has inserters), that still can discard the chest again from being eligible.

eric-anderson replied:
oh, I didn't realize those chests were a different entity. Yea, should be easy to switch over to filter for those.  I still want to ignore having inserters/loaders on them since you will but dump to those is really nice.

eric-anderson replied:
Fixed to use red-chest and blue-chest in 7cc268fab6a2342d52d60b6366cdf0138d93bca9

Gerkiz commented on comfy_panel/config.lua:
```diff
@@ -82,6 +82,9 @@ end
 
 local function spaghett()
     local spaghetti = this.gui_config.spaghett
+    if spaghetti.noop then
```
No idea what you are doing here - spaghett is not enabled unless a player admin enables it.

This can be removed.

eric-anderson replied:
If this isn't here, then the spaghett module keeps re-enabling the logistic systems technology when you pass through
https://github.com/eric-anderson/ComfyFactorio/blob/75864b385c38d2cb191d22a3bed207858d7d1179/comfy_panel/config.lua#L83

Gerkiz replied:
I'd still want to redo the trigger function instead of adding this.
But I understand now why you are adding it here.

Looks good for now.

Gerkiz commented on modules/collapse.lua:
```diff
@@ -5,8 +5,10 @@ local math_floor = math.floor
 local collapse = {
-    debug = false
+    debug = false,
+    disabled = false,
```
Same goes with this one.

eric-anderson replied:
Also reverted.

Gerkiz reviewed (changes requested):
Do fix these changes before this PR can be approved.

I don't understand your reasoning with adding new modules when your main objective was to "fix" Dungeons?

Next time, new stuff goes in separate PR.

eric-anderson committed:
Review fixes.

* Clearing corpses remains trusted only.
* Remove spurious debugging code.

eric-anderson committed:
Merge remote-tracking branch 'gh/develop' into master

eric-anderson commented:
> I don't understand your reasoning with adding new modules when your main objective was to "fix" Dungeons?

The starting "fix" was the research thing, the remainder came as I was playtesting and kept running into issues. The pull request came when I had a version that I hadn't found a problem with in several days.

> Next time, new stuff goes in separate PR.

Do you want me to reslice the commits?  I asked in the comfy dungeons channel and didn't get a response so sent the pull request in the first of the three options, it sounds like you would have preferred the third. "Let me know if you'd like to send the pull request with the full original chain of commits (each one was a checkpoint when I thought I'd achieved a goal, and in general each commit mostly worked), or if you'd like me to restructure it somehow (by file; single commit; separate commits for groups of changes, e.g. research, melee mode, changelog, balancing, bugfixes) ."

eric-anderson committed:
Update changelog module and tiered dungeon to handle move of comfy_panel to utils.gui

eric-anderson committed:
BUGFIX: code incorrectly handled floors without research

eric-anderson committed:
use type (red/blue chest) instead of position for identifying valid autostash targets for dungeons

eric-anderson committed:
Disable start_now in collapse rather than adding disable feature.

It looks like this works to keep the collapse logic from running.
It relies on the dungeons on_init function running after the collapse
one.

Gerkiz committed:
Merge branch 'develop' into master

Gerkiz commented:
Looks good. Approved the changes!

@hanakocz has the final word 👍

hanakocz reviewed (approved):
Think it is good to go now too, hopefully. :) Good job and thanks
